sphinx/builders/linkcheck.py

```diff
@@ -245,47 +235,73 @@ def write_linkstat(self, data: dict) -> None:
-            total_links = 0
-            for hyperlink in self.hyperlinks.values():
-                if self.is_ignored_uri(hyperlink.uri):
-                    self.process_result(
-                        CheckResult(hyperlink.uri, hyperlink.docname, hyperlink.lineno,
-                                    'ignored', '', 0))
-                else:
-                    self.wqueue.put(hyperlink, False)
-                    total_links += 1
-
-            done = 0
-            while done < total_links:
-                self.process_result(self.rqueue.get())
-                done += 1
+            for result in checker.check(self.hyperlinks):
+                self.process_result(result)
 
         if self._broken:
             self.app.statuscode = 1
 
+
+class HyperlinkAvailabilityChecker:
+    def __init__(self, builder: CheckExternalLinksBuilder) -> None:
+        self.builder = builder
+        self.config = builder.config
+        self.rate_limits = {}  # type: Dict[str, RateLimit]
+        self.rqueue = queue.Queue()  # type: queue.Queue
+        self.wqueue = queue.PriorityQueue()  # type: queue.PriorityQueue
+        self.workers = []  # type: List[Thread]
+
+    def invoke_threads(self) -> None:
+        for i in range(self.config.linkcheck_workers):
+            thread = HyperlinkAvailabilityCheckWorker(self.builder, self.rqueue, self.wqueue,
+                                                      self.rate_limits)
+            thread.start()
+            self.workers.append(thread)
+
+    def shutdown_threads(self) -> None:
         self.wqueue.join()
-        # Shutdown threads.
         for worker in self.workers:
             self.wqueue.put((CHECK_IMMEDIATELY, None, None, None), False)
 
+    def check(self, hyperlinks: Dict[str, Hyperlink]) -> Generator[Tuple[str, str, int, str, str, int], None, None]:  # NOQA
+        self.invoke_threads()
+
+        total_links = 0
+        for hyperlink in hyperlinks.values():
+            if self.builder.is_ignored_uri(hyperlink.uri):
+                yield CheckResult(hyperlink.uri, hyperlink.docname, hyperlink.lineno,
+                                  'ignored', '', 0)
+            else:
+                self.wqueue.put(hyperlink, False)
+                total_links += 1
+
+        done = 0
+        while done < total_links:
+            yield self.rqueue.get()
+            done += 1
+
+        self.shutdown_threads()
+
 
 class HyperlinkAvailabilityCheckWorker(Thread):
     """A worker class for checking the availability of hyperlinks."""
 
-    def __init__(self, builder: CheckExternalLinksBuilder) -> None:
+    def __init__(self, builder: CheckExternalLinksBuilder, rqueue: Queue, wqueue: Queue,
+                 rate_limits: Dict[str, RateLimit]) -> None:
         self.app = builder.app
         self.anchors_ignore = builder.anchors_ignore
         self.auth = builder.auth
         self.config = builder.config
         self.env = builder.env
```

Comment:
After some refactoring, the `builder` argument will be no longer needed. So I'll pass `config` and `env` object directly to this constructor at the final step.

Reply:
Done.